src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java

```diff
@@ -0,0 +1,407 @@
+                HttpEntity entity = httpResponse.getEntity();
+                T obj = new ObjectMapper()
+                        .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
+                        .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS, true)
+                        .configure(DeserializationFeature.ACCEPT_EMPTY_ARRAY_AS_NULL_OBJECT, true)
+                        .configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true)
+                        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+                        .readValue(IOUtils.toString(entity.getContent(), StandardCharsets.UTF_8), type);
+
+                LOGGER.info(type + " received");
+
+                return obj;
+            } else {
+                throw new IOException("No response reveived");
+            }
+        }
+    }
+
+    private void sendPatch(String apiUrl, String body, String authorizationHeader) throws IOException, MalformedURLException, URISyntaxException {
+        apiUrl = encodeURI(apiUrl);
+        LOGGER.trace(String.format("sendPatch: URL: %s ", apiUrl));
+        LOGGER.trace(String.format("sendPatch: BODY: %s ", body));
+        HttpPatch httpPatch = new HttpPatch(apiUrl);
+        httpPatch.addHeader("Accept", "application/json");
+        httpPatch.addHeader("Content-Type", "application/json; charset=utf-8");
+        httpPatch.addHeader("Authorization", authorizationHeader);
+        StringEntity entity = new StringEntity(body, StandardCharsets.UTF_8);
+        httpPatch.setEntity(entity);
+
+        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+            HttpResponse httpResponse = httpClient.execute(httpPatch);
+            if (null != httpResponse && httpResponse.getStatusLine().getStatusCode() != 200) {
+                LOGGER.error("sendPatch: " + httpResponse.toString());
+                LOGGER.error("sendPatch: " + EntityUtils.toString(httpResponse.getEntity(), StandardCharsets.UTF_8));
+                throw new IllegalStateException("An error was returned in the response from the Azure DevOps server. See the previous log messages for details");
+            } else if (null != httpResponse) {
+                LOGGER.debug("sendPatch: " + httpResponse.toString());
+                LOGGER.info("sendPatch: Patch success!");
+            }
+        }
+    }
+
+    private static Optional<String> ensureTrailingSlash(Optional<String> uri) {
+        if (uri != null && uri.isPresent()) {
+            return Optional.of(uri.get().endsWith("/") ? uri.get() : uri.get() + "/");
+        }
+        return Optional.empty();
+    }
+
+    private static String encodeURI(String uri) throws MalformedURLException, URISyntaxException {
+        URL url = new URL(uri);
+        return (new URI(url.getProtocol(), url.getUserInfo(), url.getHost(), url.getPort(), url.getPath(), url.getQuery(), url.getRef())).toString();
+    }
```

Comment:
I'm not clear on what this method is actually trying to do, but `String` -> `URL` -> `URI` -> `String` seems an odd conversion

Reply:
The initial string could potentially have spaces in the URL.  java.net.URLEncoder encodes spaces to '+' which don't work in the address portion of the URI, only in form data.  This method gets around that by initializing an instance of java.net.URL then converting to a true, encoded URI.

I could just do a replace on ' ' to '%20', but this seemed like a better method that handles both address and querystring unless you have another way.